api/handlers.go

```diff
@@ -165,3 +171,404 @@ func (api *RecipeAPI) AddRecipeHandler(w http.ResponseWriter, req *http.Request)
+		return
+	}
+	w.Header().Set("content-type", "application/json")
+	w.Write(output)
+}
+
+//UpdateInstanceHandler - update specific recipe by ID
+func (api *RecipeAPI) UpdateInstanceHandler(w http.ResponseWriter, req *http.Request) {
+	ctx := req.Context()
+
+	//Get Update Recipe ID
+	vars := mux.Vars(req)
+	recipeID := vars["id"]
+	instanceID := vars["instance_id"]
+	logD := log.Data{"recipe_id": recipeID, "instance_id": instanceID}
+
+	// Read body
+	b, err := ioutil.ReadAll(req.Body)
+	defer req.Body.Close()
+	if err != nil {
+		log.Event(ctx, "error in reading request body", log.ERROR, log.Error(err))
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	// Unmarshal to the shape of Instance struct
+	var updates recipe.Instance
+	err = json.Unmarshal(b, &updates)
+	if err != nil {
+		log.Event(ctx, "error returned from json unmarshal", log.ERROR, log.Error(err))
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	// Validation to check fields of instance and if all the codelists of the instance are entered if update of codelist given
+	err = updates.ValidateUpdateInstance()
+	if err != nil {
+		log.Event(ctx, "bad request error for invalid updates given in request body", log.ERROR, log.Error(err))
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	// Find index of specific interface in output_instances of the recipe
+	currentRecipe, err := api.dataStore.Backend.GetRecipe(recipeID)
+	if err != nil {
+		log.Event(ctx, "error retrieving specific recipe from mongo", log.ERROR, log.Error(err), logD)
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	instanceIndex := -1
+	for i, instance := range currentRecipe.OutputInstances {
+		if instance.DatasetID == instanceID {
+			instanceIndex = i
+			break
+		}
+	}
+	if instanceIndex == -1 {
+		log.Event(ctx, "error retrieving specific instance of recipe from mongo", log.ERROR, logD)
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	// Set the non-updated fields to values of the currentRecipe - this needs to be done otherwise all fields in array will be overwritten
+	updates.DatasetID = instanceID
+	if updates.Editions == nil || len(updates.Editions) == 0 {
+		updates.Editions = currentRecipe.OutputInstances[instanceIndex].Editions
+	}
+	if updates.Title == "" {
+		updates.Title = currentRecipe.OutputInstances[instanceIndex].Title
+	}
+	if updates.CodeLists == nil || len(updates.CodeLists) == 0 {
+		updates.CodeLists = currentRecipe.OutputInstances[instanceIndex].CodeLists
+	}
+
+	// Update Recipe to Mongo
+	update := bson.M{"$set": bson.M{"output_instances." + strconv.Itoa(instanceIndex): updates}}
+	err = api.dataStore.Backend.UpdateRecipe(recipeID, update)
+	if err != nil {
+		log.Event(ctx, "error updating recipe to mongo", log.ERROR, log.Error(err), logD)
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	// Marshal to output the newly updated recipe
+	updatedRecipe, _ := api.dataStore.Backend.GetRecipe(recipeID)
```

Comment:
Always handle the errors please

Reply:
Done in commit "move bson statements to mongo package - mongo package encapsulation" as GetRecipe in this handler is removed in which no error needs to be handled on this section